### ISSUE TITLE
colcon doesnt aggregate cmake args

### ISSUE THREAD

**Issue by mikaelarguedas** [colcon/colcon-core#18]
`colcon build --symlink-install --cmake-args ' -DSECURITY=1 -DBUILD_TESTING=1' --event-handler console_cohesion+ --build-base $ROS2_WS/build_debug_isolated --install-base $ROS2_WS/install_debug_isolated --cmake-args ' -DCMAKE_BUILD_TYPE=Debug'`

Only Debug is taken into account. DSECURITY and DBUILD_TESTING are ignored. This would prevent us from passing CMake arguments on CI

**Comment by dirk-thomas** [comment]
Each CMake argument needs to be quoted separately. Please try:

```
colcon build --symlink-install --cmake-args ' -DSECURITY=1' ' -DBUILD_TESTING=1' --event-handler console_cohesion+ --build-base $ROS2_WS/build_debug_isolated --install-base $ROS2_WS/install_debug_isolated --cmake-args ' -DCMAKE_BUILD_TYPE=Debug'

**Comment by mikaelarguedas** [comment]
must be user error then...

I tried quoting them separately and also adding `--cmake-args` before each cmake argument or place them in different parts of the invocation, but I'm still unable to build `test_security` with SECURITY enabled.

I'm using this branch : https://github.com/ros2/system_tests/tree/is_security_on

this invocation: colcon build --symlink-install --event-handler console_direct+ --build-base $ROS2_WS/build_debug_isolated --install-base $ROS2_WS/install_debug_isolated --packages-select test_security --cmake-args ' -DCMAKE_BUILD_TYPE=Debug' --cmake-args ' -DSECURITY=1' --cmake-args ' -DBUILD_TESTING=1'

and this is my stderr output:
```
--- stderr: test_security
CMake Warning at CMakeLists.txt:26 (message):
  SECURITY is not ON


CMake Warning at CMakeLists.txt:30 (message):
  BUILD_TESTING is ON


CMake Warning at CMakeLists.txt:293 (message):
  SECURITY not ON


---
```

I'm using the latest version of colcon from source

**Comment by mikaelarguedas** [comment]
Looks like the previous linux build also didnt run the security tests: https://ci.ros2.org/view/colcon/job/colcon_ci_linux/9: "test_security [2.12s]"
And the list of tests shows only the linters.

I just reran one with only that package and it seems to have worked. Any idea what the difference is ?
https://ci.ros2.org/view/colcon/job/colcon_ci_linux/10
" Finished <<< test_security [270.51s]"

**Comment by mikaelarguedas** [comment]
This seems tied to the position of the arguments:

This doesn't work (cmake doesn't get the -DSECURITY)
```
colcon build --base-paths "src" --build-base "build" --install-base "install" --cmake-args " -DBUILD_TESTING=1" --cmake-args " -DSECURITY=ON" --event-handler console_cohesion+ --packages-select test_security  --cmake-args " -DCMAKE_BUILD_TYPE=Release"
```

This works:
```
colcon build --base-paths "src" --build-base "build" --install-base "install"  --cmake-args " -DCMAKE_BUILD_TYPE=Release" --cmake-args " -DBUILD_TESTING=1" --cmake-args " -DSECURITY=ON" --event-handler console_cohesion+ --packages-select test_security
```

**Comment by dirk-thomas** [comment]
Passing multiple `--cmake-args` is not supported by `argparse` natively. It is trivial to implement a custom `ExtendAction` which enables this.

The question is if it is worth to do so with the chance of the user getting confused why it works or expects it to work on all arguments (which would only be the case if all of them would use that custom action).

On the other hand we could update the CI logic to avoid adding a second `--cmake-args` when one is already passed and insert the arguments in the right location. The user would still only be allowed to use `--cmake-args` once.

To follow the KISS principle I am leaning toward the second option.

**Comment by dirk-thomas** [comment]
> On the other hand we could update the CI logic to avoid adding a second `--cmake-args` when one is already passed and insert the arguments in the right location. The user would still only be allowed to use `--cmake-args` once.

This is implemented https://github.com/ros2/ci/pull/132/commits/3ec23d3bbc2db3a4ca1086f1c66842d13aacc74f as part of ros2/ci#132. The `--cmake-args` / `--ament-cmake-args` are now being generated without repeating the "keyword". Even when the Job parameters come in they get merged into the other "hard coded" options without duplicating the "keyword".

**Comment by mikaelarguedas** [comment]
cool :+1: 

So now the only restriction on CI is to pass 1 and only 1 occurrence of "--cmake-args"

And for users using colcon locally they need to ensure to pass "--cmake-args" only once in the entire invocation.

Is that correct ?

**Comment by dirk-thomas** [comment]
> Is that correct ?

:+1:
